controllers/apps/components/util/component_switchover_util.go

```diff
@@ -0,0 +1,628 @@
+func HandleSwitchover(ctx context.Context,
+	cli client.Client,
+	cluster *appsv1alpha1.Cluster,
+	component *intctrlcomputil.SynthesizedComponent,
+	obj client.Object) error {
+	if component.SwitchoverCandidate == nil {
+		return nil
+	}
+	// check if all Pods have role label
+	podList, err := GetRunningPods(ctx, cli, obj)
+	if err != nil {
+		return err
+	}
+	// if the switchover candidate index is invalid, we do nothing.
+	if component.SwitchoverCandidate.Index > int32(len(podList)-1) || component.SwitchoverCandidate.Index < 0 {
+		return nil
+	}
+	for _, pod := range podList {
+		// if the pod does not have the role label, we do nothing.
+		if v, ok := pod.Labels[constant.RoleLabelKey]; !ok || v == "" {
+			return nil
+		}
+	}
+
+	// check if the switchover is needed
+	needSwitchover, err := NeedDealWithSwitchover(ctx, cli, cluster, component)
+	if err != nil {
+		return err
+	}
+	if needSwitchover {
+		// create a job to do switchover and check the result
+		if err := DoSwitchover(ctx, cli, cluster, component); err != nil {
+			return err
+		}
+	} else {
+		// if the switchover is not needed, it means that the switchover has been completed, and the switchover job can be deleted.
+		if err := PostOpsSwitchover(ctx, cli, cluster, component); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// HandleRoleSync synchronizes the results of roleChange to switchoverCandidate if necessary.
+func HandleRoleSync(ctx context.Context,
+	cli client.Client,
+	cluster *appsv1alpha1.Cluster,
+	component *intctrlcomputil.SynthesizedComponent) error {
+	if component.SwitchoverCandidate == nil {
+		return nil
+	}
+	// if the failover sync is not enabled, we do not sync the failover result.
+	if !component.SwitchoverCandidate.RoleSync {
+		return nil
+	}
+	switchoverCondition := meta.FindStatusCondition(cluster.Status.Conditions, appsv1alpha1.ConditionTypeSwitchoverPrefix+component.Name)
+	// if the switchover condition is exist and status is not true, it means that the switchover is not completed, we do not sync the failover result.
+	if switchoverCondition != nil && switchoverCondition.Status != metav1.ConditionTrue {
+		return nil
+	}
+	ok, pod, err := checkPodRoleLabelConsistency(ctx, cli, cluster, component)
+	if err != nil {
+		return err
+	}
+	if ok || pod == nil {
+		return nil
+	}
+	// synchronize the current primary or leader pod ordinal to switchoverCandidate.Index and set switchoverCandidate.Operator to equal.
+	_, o := ParseParentNameAndOrdinal(pod.Name)
+	for index, compSpec := range cluster.Spec.ComponentSpecs {
+		if compSpec.Name == component.Name {
+			cluster.Spec.ComponentSpecs[index].SwitchoverCandidate = &appsv1alpha1.SwitchoverCandidate{
+				Index:    o,
+				Operator: appsv1alpha1.CandidateOpEqual,
+				RoleSync: true,
+			}
+		}
+	}
+	return nil
+}
+
+// NeedDealWithSwitchover checks whether we need to handle the switchover process.
+func NeedDealWithSwitchover(ctx context.Context,
```

Comment:
NeedDealWithSwitchover -> needDealWithSwitchover

Reply:
> NeedDealWithSwitchover -> needDealWithSwitchover

The lifeCycle of the previous caller is in another package, I will fix it